crates/iceberg/src/arrow/reader.rs

```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+        PrimitiveLiteral::Float(value) => Ok(Box::new(Float32Array::new_scalar(value.as_f32()))),
+        PrimitiveLiteral::Double(value) => Ok(Box::new(Float64Array::new_scalar(value.as_f64()))),
+        l => Err(Error::new(
+            ErrorKind::DataInvalid,
+            format!("Unsupported literal type: {:?}", l),
+        )),
+    }
+}
+
+/// Recursively get the leaf column from the record batch. Assume that the nested columns in
+/// struct is projected to a single column.
+fn get_leaf_column(column: &ArrayRef) -> std::result::Result<ArrayRef, ArrowError> {
+    match column.data_type() {
+        DataType::Struct(fields) => {
+            if fields.len() != 1 {
+                return Err(ArrowError::SchemaError(
+                    "Struct column should have only one field after projection"
+                        .parse()
+                        .unwrap(),
+                ));
+            }
+            let struct_array = column.as_any().downcast_ref::<StructArray>().unwrap();
+            get_leaf_column(struct_array.column(0))
+        }
+        _ => Ok(column.clone()),
+    }
+}
+
+impl<'a> BoundPredicateVisitor for PredicateConverter<'a> {
+    type T = Box<dyn ArrowPredicate>;
+
+    fn always_true(&mut self) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            |batch| Ok(BooleanArray::from(vec![true; batch.num_rows()])),
+        )))
+    }
+
+    fn always_false(&mut self) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            |batch| Ok(BooleanArray::from(vec![false; batch.num_rows()])),
+        )))
+    }
+
+    fn and(&mut self, mut lhs: Self::T, mut rhs: Self::T) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            move |batch| {
+                let left = lhs.evaluate(batch.clone())?;
+                let right = rhs.evaluate(batch)?;
+                and(&left, &right)
+            },
+        )))
+    }
+
+    fn or(&mut self, mut lhs: Self::T, mut rhs: Self::T) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            move |batch| {
+                let left = lhs.evaluate(batch.clone())?;
+                let right = rhs.evaluate(batch)?;
+                or(&left, &right)
+            },
+        )))
+    }
+
+    fn not(&mut self, mut inner: Self::T) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            move |batch| {
+                let pred_ret = inner.evaluate(batch)?;
+                not(&pred_ret)
+            },
+        )))
+    }
+
+    fn is_null(
+        &mut self,
+        reference: &BoundReference,
+        _predicate: &BoundPredicate,
+    ) -> Result<Self::T> {
+        let projected_mask = self.bound_reference(reference)?;
```

Comment:
This seems incorrect to me. Let's say the predicate is `a is null AND b >1`, then the batch passed to this `ArrowPredicateFn` is constructed by projection mask of [a, b]. I think one possible solution is to use same project mask for all predicates, and pass the column_idx to get_leaf_column.

Reply:
Hmm, as `RowFilter` API design accepts multiple `ArrowPredicate`s. Each `ArrowPredicate` has its `projection` and the API doc said the API will be passed batches that contains the columns specified in `projection`. I think it should be projected for each `ArrowPredicate` and `projection`.

Reply:
I will go to test it in arrow-rs to verify it.

Reply:
Ah, no, although I used multiple `ArrowPredicates` before, it was changed to one `ArrowPredicates` after you suggested. So now we generate only one `ArrowPredicates`.

Hmm, let me think how do deal with it.

Reply:
Based on what I searched and the kindly reply on the issue, I think there is no way to do nested projection on RecordBatch currently.

To implement the feature in arrow-rs might block this. I tend to finish top-level column only in this PR.

WDYT, @liurenjie1024 ?